### PR TITLE
spd: dispatcher for interacting with the Trusty TEE

### PR DESCRIPTION
**Pull request by vwadekar**
This patch adds the secure payload dispatcher for interacting
with Google's Trusty TEE. Documentation for Trusty can be found
at https://source.android.com/security/trusty

Original authors:
-----------------
* Arve Hjønnevåg <arve@android.com>
* Michael Ryleev <gmar@google.com>

Signed-off-by: Varun Wadekar <vwadekar@nvidia.com>

### CODE REVIEW

**Comment by vwadekar** [comment]
This is the base change which I would like to get reviewed and merged first. I will push additional changes from our downstream tree as another PR.

**Comment by danh-arm** [comment]
jenkins:test this please

File: services/spd/trusty/smcall.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+#define	SMC_ENTITY_STD			4	/* Standard Service calls */
+#define	SMC_ENTITY_RESERVED		5	/* Reserved for future use */
+#define	SMC_ENTITY_TRUSTED_APP		48	/* Trusted Application calls */
+#define	SMC_ENTITY_TRUSTED_OS		50	/* Trusted OS calls */
+#define SMC_ENTITY_LOGGING              51	/* Used for secure -> nonsecure logging */
+#define	SMC_ENTITY_SECURE_MONITOR	60	/* Trusted OS calls internal to secure monitor */
+
+/* FC = Fast call, SC = Standard call */
+#define SMC_SC_RESTART_LAST	SMC_STDCALL_NR  (SMC_ENTITY_SECURE_MONITOR, 0)
+#define SMC_SC_NOP		SMC_STDCALL_NR  (SMC_ENTITY_SECURE_MONITOR, 1)
```
Comment (danh-arm) [inline services/spd/trusty/smcall.h]: non-gating comment: Some of the defintions here and in sm_err.h should probably be in smcc.h (or used directly from there if already defined).

File: services/spd/trusty/trusty.c [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+	bl31_register_bl32_init(trusty_init);
+
+	psci_register_spd_pm_hook(&trusty_pm);
+
+	flags = 0;
+	set_interrupt_rm_flag(flags, NON_SECURE);
+	ret = register_interrupt_type_handler(INTR_TYPE_S_EL1,
+					      trusty_fiq_handler,
+					      flags);
```
Comment (danh-arm) [inline services/spd/trusty/trusty.c]: non-gating comment: We think this interrupt model is going to be problematic in GICv3 systems but that can be addressed later.

**Comment by vwadekar** [comment]
Thanks Dan.

**Pull request merged**